#Importaciones

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import datetime
from datetime import datetime
import re

In [2]:
from google.colab import drive
drive.mount('/content/drive') 

Mounted at /content/drive


In [3]:
db_train=pd.read_csv('/content/drive/MyDrive/DATASCIENCE/Practico/Practico 3/train (3).csv')

In [4]:
metadata=pd.read_csv('/content/drive/MyDrive/DATASCIENCE/Practico/Practico 3/metadata.csv',sep=';')

#Depuracion de datos

##metadata

In [5]:
metadata.isnull().sum()

asset_id                  0
content_id               21
title                     0
reduced_title             0
episode_title          4147
show_type                 4
released_year             0
country_of_origin         4
category                  0
keywords                  2
description               2
reduced_desc              0
cast_first_name        8732
credits_first_name    12554
run_time_min              0
audience                  1
made_for_tv               0
close_caption             0
sex_rating                0
violence_rating           0
language_rating           0
dialog_rating             0
fv_rating                 0
pay_per_view              0
pack_premium_1            0
pack_premium_2            0
create_date               0
modify_date               0
start_vod_date            0
end_vod_date              0
dtype: int64

###content_id

In [6]:
metadata[metadata['content_id'].isnull()]

asset_id  content_id                   title        reduced_title  \
1540      15032         NaN     Audios y Subtítulos   Tuto_Audio_Subtitu   
1541      22963         NaN        Control Parental   Tuto_Control_Paren   
1542      32637         NaN          Control Remoto   Tuto_Control_Remot   
1543      21822         NaN               Favoritos        Tuto_Favorito   
1544      17157         NaN  Grabaciones en la Nube   Tuto_Grab_En_La_Nu   
1545      24583         NaN      Múltiples Perfiles   Tuto_Multi_Perfile   
1546      27036         NaN    Programas Anteriores   Tuto_Prog_Anterior   
18197      2046         NaN     Audios y Subtítulos  Tuto_Audio_Subti_Py   
18198      1105         NaN        Control Parental  Tuto_Control_Par_Py   
18199     28435         NaN          Control Remoto  Tuto_Control_Rem_Py   
18200     17563         NaN               Favoritos     Tuto_Favorito_Py   
18201     18599         NaN  Grabaciones en la Nube   Tuto_Grab_En_La_Py   
18202     27733         NaN      Múltiples Perfiles  Tuto_Multi_Perfi_Py   
18203     25592         NaN    Programas Anteriores  Tuto_Prog_Anteri_Py   
22895     15861         NaN     Audios y Subtítulos   Tuto_Audio_Subtitu   
22896      3938         NaN        Control Parental   Tuto_Control_Paren   
22897     12983         NaN          Control Remoto   Tuto_Control_Remot   
22898     16881         NaN               Favoritos        Tuto_Favorito   
22899     19462         NaN  Grabaciones en la Nube   Tuto_Grab_En_La_Nu   
22900     28382         NaN      Múltiples Perfiles   Tuto_Multi_Perfile   
22901      6893         NaN    Programas Anteriores   Tuto_Prog_Anterior   

      episode_title show_type  released_year country_of_origin  \
1540            NaN  Tutorial         2019.0                AR   
1541            NaN  Tutorial         2019.0                AR   
1542            NaN  Tutorial         2019.0                AR   
1543            NaN  Tutorial         2019.0                AR   
1544            NaN  Tutorial         2019.0                AR   
1545            NaN  Tutorial         2019.0                AR   
1546            NaN  Tutorial         2019.0                AR   
18197           NaN  Tutorial         2019.0                AR   
18198           NaN  Tutorial         2019.0                AR   
18199           NaN  Tutorial         2019.0                AR   
18200           NaN  Tutorial         2019.0                AR   
18201           NaN  Tutorial         2019.0                AR   
18202           NaN  Tutorial         2019.0                AR   
18203           NaN  Tutorial         2019.0                AR   
22895           NaN  Tutorial         2019.0                AR   
22896           NaN  Tutorial         2019.0                AR   
22897           NaN  Tutorial         2019.0                AR   
22898           NaN  Tutorial         2019.0                AR   
22899           NaN  Tutorial         2019.0                AR   
22900           NaN  Tutorial         2019.0                AR   
22901           NaN  Tutorial         2019.0                AR   

                            category    keywords  ... language_rating  \
1540   Tutoriales/Decodificador Flow  Tutoriales  ...               N   
1541   Tutoriales/Decodificador Flow  Tutoriales  ...               N   
1542   Tutoriales/Decodificador Flow  Tutoriales  ...               N   
1543   Tutoriales/Decodificador Flow  Tutoriales  ...               N   
1544   Tutoriales/Decodificador Flow  Tutoriales  ...               N   
1545   Tutoriales/Decodificador Flow  Tutoriales  ...               N   
1546   Tutoriales/Decodificador Flow  Tutoriales  ...               N   
18197  Tutoriales/Decodificador Flow  Tutoriales  ...               N   
18198  Tutoriales/Decodificador Flow  Tutoriales  ...               N   
18199  Tutoriales/Decodificador Flow  Tutoriales  ...               N   
18200  Tutoriales/Decodificador Flow  Tutoriales  ...               N   
18201

Se puede apreciar que estos casos son tutoriales/decodificador flow, por lo que no consideramos que sean contenidos recomendables y seran quitados.

In [7]:
metadata=metadata.drop(metadata[metadata['category']=='Tutoriales/Decodificador Flow'].index,axis=0)

In [8]:
metadata.shape

(33123, 30)

In [9]:
metadata.isnull().sum()

asset_id                  0
content_id                0
title                     0
reduced_title             0
episode_title          4126
show_type                 4
released_year             0
country_of_origin         4
category                  0
keywords                  2
description               2
reduced_desc              0
cast_first_name        8711
credits_first_name    12533
run_time_min              0
audience                  1
made_for_tv               0
close_caption             0
sex_rating                0
violence_rating           0
language_rating           0
dialog_rating             0
fv_rating                 0
pay_per_view              0
pack_premium_1            0
pack_premium_2            0
create_date               0
modify_date               0
start_vod_date            0
end_vod_date              0
dtype: int64

###episode_title

In [10]:
metadata[metadata['episode_title'].isnull()]

asset_id  content_id                                  title  \
1         24940         1.0                                7 Cajas   
2         21939         2.0  La Maldición de las Hormigas Gigantes   
3          9005         3.0                   Una Mujer Fantástica   
4          7391         4.0                              Star Trek   
10        21684         9.0             Star Trek: En la oscuridad   
...         ...         ...                                    ...   
33118     11656      4367.0                     Espejito, espejito   
33123     11482      4368.0                     15:17 Tren a París   
33124     11218      4369.0                       Grito de batalla   
33132     27690      4370.0              2047: Vistas de la muerte   
33137     19624      4371.0       Zapatos rojos y los siete enanos   

             reduced_title episode_title show_type  released_year  \
1                  7_Cajas           NaN  Película         2012.0   
2      La_Maldicion_de_las           NaN  Película         2016.0   
3      Una_Mujer_Fantastic           NaN  Película         2017.0   
4                Star_Trek           NaN  Película         2009.0   
10     Star_Trek__En_la_os           NaN  Película         2013.0   
...                    ...           ...       ...            ...   
33118    Espejito_Espejito           NaN  Película         2012.0   
33123   15_17_Tren_a_Paris           NaN  Película         2018.0   
33124     Grito_de_batalla           NaN  Película         1963.0   
33132  2047__Vistas_de_la_           NaN  Película         2014.0   
33137  Zapatos_rojos_y_los           NaN  Película         2019.0   

      country_of_origin                  category  \
1                    PY           Suspenso/Acción   
2                    FI            Terror/Comedia   
3                    CL                     Drama   
4                    US  Ciencia Ficción/Aventura   
10                   US  Ciencia Ficción/Aventura   
...                 ...                       ...   
33118                US         Aventura/Fantasía   
33123                US            Drama/Suspenso   
33124                US              Drama/Acción   
33132                IT    Acción/Ciencia Ficción   
33137                US        Infantil/Animación   

                                                keywords  ... language_rating  \
1                 Latinoamérica,Pobreza,Crimen,Pandillas  ...               N   
2      Criaturas,Plagas,Adolescentes,Fantasía,Video J...  ...               N   
3                             LGBT,Mujeres,Latinoamérica  ...               N   
4            Fantasía,Galaxia,Futurismo,Aliens,Criaturas  ...               N   
10      Fantasía,Galaxia,Futurismo,Aliens,Criaturas,Saga  ...               N   
...                                                  ...  ...             ...   
33118                                Princesas,De Libros  ...               N   
33123  Hechos Reales,Terroristas,Viajes,Amigos,Militares  ...               N   
33124      Clásicas,BYN,Guerra Mundial,Familia,Militares  ...               N   
33132                                 Futurismo,Distopía  ...               N   
33137               Aventuras,Criaturas,Princesas,Brujas  ...               N   

      dialog_rating fv_rating pay_per_view  pack_premium_1 pack_premium_2  \
1                 N         N            Y               N              N   
2                 N         N            N               N              N   
3                 N         N            N               Y              N   
4                 N         N            Y               N              N   
10                N         N            Y               N              N   
...             ...       ...          ...             ...            ...   
33118             N         N            N               N              N   
33123             N         N            N               N              N   
33124             N         N     

para este caso simplemente vamos a copiar el titulo reduced_title

In [11]:
metadata.loc[metadata['episode_title'].isnull(),'episode_title']=metadata[metadata['episode_title'].isnull()]['reduced_title']

In [12]:
metadata.head()

asset_id  content_id                                  title  \
0     15188         0.0              Ep:17 Tiempos Compulsivos   
1     24940         1.0                                7 Cajas   
2     21939         2.0  La Maldición de las Hormigas Gigantes   
3      9005         3.0                   Una Mujer Fantástica   
4      7391         4.0                              Star Trek   

         reduced_title        episode_title show_type  released_year  \
0   Tiempos_Compul_E17          Episodio 17     Serie         2012.0   
1              7_Cajas              7_Cajas  Película         2012.0   
2  La_Maldicion_de_las  La_Maldicion_de_las  Película         2016.0   
3  Una_Mujer_Fantastic  Una_Mujer_Fantastic  Película         2017.0   
4            Star_Trek            Star_Trek  Película         2009.0   

  country_of_origin                  category  \
0                AR                     Drama   
1                PY           Suspenso/Acción   
2                FI            Terror/Comedia   
3                CL                     Drama   
4                US  Ciencia Ficción/Aventura   

                                            keywords  ... language_rating  \
0  Trastornos,Médicos,Tragicómica,Telenovela,Enfe...  ...               N   
1             Latinoamérica,Pobreza,Crimen,Pandillas  ...               N   
2  Criaturas,Plagas,Adolescentes,Fantasía,Video J...  ...               N   
3                         LGBT,Mujeres,Latinoamérica  ...               N   
4        Fantasía,Galaxia,Futurismo,Aliens,Criaturas  ...               N   

  dialog_rating fv_rating pay_per_view  pack_premium_1 pack_premium_2  \
0             N         N            N               N              N   
1             N         N            Y               N              N   
2             N         N            N               N              N   
3             N         N            N               Y              N   
4             N         N            Y               N              N   

              create_date             modify_date          start_vod_date  \
0  2017-12-01T10:18:15.0Z  2019-01-26T06:37:18.0Z  2017-12-01T00:00:00.0Z   
1  2017-12-19T20:58:15.0Z  2019-09-17T19:02:03.0Z  2017-12-15T00:00:00.0Z   
2  2018-02-16T13:51:07.0Z  2020-04-28T14:16:38.0Z  2018-01-25T00:00:00.0Z   
3  2018-05-26T11:58:44.0Z  2019-11-15T03:00:23.0Z  2018-05-27T00:00:00.0Z   
4  2019-05-03T20:07:24.0Z  2020-04-09T04:37:29.0Z  2019-05-02T00:00:00.0Z   

             end_vod_date  
0  2020-12-01T23:59:59.0Z  
1  2022-12-14T23:59:59.0Z  
2  2020-12-01T23:59:59.0Z  
3  2021-04-30T23:59:59.0Z  
4  2020-12-31T23:59:59.0Z  

[5 rows x 30 columns]

In [13]:
metadata.isnull().sum()

asset_id                  0
content_id                0
title                     0
reduced_title             0
episode_title             0
show_type                 4
released_year             0
country_of_origin         4
category                  0
keywords                  2
description               2
reduced_desc              0
cast_first_name        8711
credits_first_name    12533
run_time_min              0
audience                  1
made_for_tv               0
close_caption             0
sex_rating                0
violence_rating           0
language_rating           0
dialog_rating             0
fv_rating                 0
pay_per_view              0
pack_premium_1            0
pack_premium_2            0
create_date               0
modify_date               0
start_vod_date            0
end_vod_date              0
dtype: int64

###show_type

In [14]:
metadata[metadata['show_type'].isnull()]

asset_id  content_id                            title  \
3187      14004         9.0       Star Trek: En la oscuridad   
6421      11154       948.0          Un verano en pantalones   
7674      27728         9.0       Star Trek: En la oscuridad   
32927     24509      4315.0  El día que el deporte se detuvo   

             reduced_title        episode_title show_type  released_year  \
3187   Star_Trek__En_la_os  Star_Trek__En_la_os       NaN         2013.0   
6421   Un_Verano_En_Pantal  Un_Verano_En_Pantal       NaN         2005.0   
7674   Star_Trek_En_la_DOB  Star_Trek_En_la_DOB       NaN         2013.0   
32927  El_dia_que_el_depor  El_dia_que_el_depor       NaN         2021.0   

      country_of_origin                category  \
3187                 US  Acción/Ciencia Ficción   
6421                 US                 Comedia   
7674                 US  Acción/Ciencia Ficción   
32927                US      Documental/Deporte   

                                                keywords  ... language_rating  \
3187   Aventuras,Saga,Futurismo,Galaxia,Criaturas,Ali...  ...               N   
6421                     Mujeres,De Libros,Amigos,Viajes  ...               N   
7674   Aventuras,Saga,Futurismo,Galaxia,Criaturas,Ali...  ...               N   
32927            Pandemias,Deportistas,De Color,Sociedad  ...               N   

      dialog_rating fv_rating pay_per_view  pack_premium_1 pack_premium_2  \
3187              N         N            N               N              N   
6421              N         N            N               N              N   
7674              N         N            N               N              N   
32927             N         N            N               N              Y   

                  create_date             modify_date          start_vod_date  \
3187   2020-11-16T21:16:01.0Z  2020-12-09T17:10:20.0Z  2020-11-20T00:00:00.0Z   
6421   2020-12-29T18:10:46.0Z  2020-12-29T20:06:05.0Z  2020-12-02T00:00:00.0Z   
7674   2020-12-11T15:09:21.0Z  2020-12-24T15:06:09.0Z  2020-12-15T00:00:00.0Z   
32927  2021-03-29T16:16:36.0Z  2021-03-29T17:06:04.0Z  2021-03-24T00:00:00.0Z   

                 end_vod_date  
3187   2020-12-14T23:59:59.0Z  
6421   2021-01-01T23:59:00.0Z  
7674   2022-11-20T23:59:00.0Z  
32927  2022-01-10T23:59:00.0Z  

[4 rows x 30 columns]

In [15]:
metadata[metadata['content_id']==9.0]

asset_id  content_id                       title        reduced_title  \
10        21684         9.0  Star Trek: En la oscuridad  Star_Trek__En_la_os   
24        28703         9.0  Star Trek: En la oscuridad  Star_Trek__En_la_os   
3187      14004         9.0  Star Trek: En la oscuridad  Star_Trek__En_la_os   
3331      14049         9.0  Star Trek: En la oscuridad  Star_Trek__En_la_os   
7673      32237         9.0  Star Trek: En la oscuridad  Star_Trek__En_la_os   
7674      27728         9.0  Star Trek: En la oscuridad  Star_Trek_En_la_DOB   
29046     13209         9.0  Star Trek: En la oscuridad  Star_Trek__En_la_os   
29047     15413         9.0  Star Trek: En la oscuridad  Star_Trek__En_la_os   

             episode_title show_type  released_year country_of_origin  \
10     Star_Trek__En_la_os  Película         2013.0                US   
24     Star_Trek__En_la_os  Película         2013.0                US   
3187   Star_Trek__En_la_os       NaN         2013.0                US   
3331   Star_Trek__En_la_os  Película         2013.0                US   
7673   Star_Trek__En_la_os  Película         2013.0                US   
7674   Star_Trek_En_la_DOB       NaN         2013.0                US   
29046  Star_Trek__En_la_os  Película         2013.0                US   
29047  Star_Trek__En_la_os  Película         2013.0                US   

                       category  \
10     Ciencia Ficción/Aventura   
24     Ciencia Ficción/Aventura   
3187     Acción/Ciencia Ficción   
3331     Acción/Ciencia Ficción   
7673     Acción/Ciencia Ficción   
7674     Acción/Ciencia Ficción   
29046    Acción/Ciencia Ficción   
29047    Acción/Ciencia Ficción   

                                                keywords  ... language_rating  \
10      Fantasía,Galaxia,Futurismo,Aliens,Criaturas,Saga  ...               N   
24      Fantasía,Galaxia,Futurismo,Aliens,Criaturas,Saga  ...               N   
3187   Aventuras,Saga,Futurismo,Galaxia,Criaturas,Ali...  ...               N   
3331   Aventuras,Saga,Futurismo,Galaxia,Criaturas,Ali...  ...               N   
7673   Aventuras,Saga,Futurismo,Galaxia,Criaturas,Ali...  ...               N   
7674   Aventuras,Saga,Futurismo,Galaxia,Criaturas,Ali...  ...               N   
29046                      Saga,Futurismo,Galaxia,Aliens  ...               N   
29047                      Saga,Futurismo,Galaxia,Aliens  ...               N   

      dialog_rating fv_rating pay_per_view  pack_premium_1 pack_premium_2  \
10                N         N            Y               N              N   
24                N         N            N               Y              N   
3187              N         N            N               N              N   
3331              N         N            N               N              N   
7673              N         N            N               N              N   
7674              N         N            N               N              N   
29046             N         N            N               N              N   
29047             N         N            N               N              N   

                  create_date             modify_date          start_vod_date  \
10     2020-03-06T18:21:59.0Z  2020-11-28T07:07:45.0Z  2020-01-03T00:00:00.0Z   
24     2020-07-21T14:49:55.0Z  2020-07-21T17:06:10.0Z  2020-07-04T00:00:00.0Z   
3187   2020-11-16T21:16:01.0Z  2020-12-09T17:10:20.0Z  2020-11-20T00:00:00.0Z   
3331   2020-12-08T14:47:32.0Z  2020-12-09T17:09:15.0Z  2020-11-20T00:00:00.0Z   
7673   2020-12-11T14:59:08.0Z  2020-12-24T15:06:09.0Z  2020-12-15T00:00:00.0Z   
7674   2020-12-11T15:09:21.0Z  2020-12-24T15:06:09.0Z  2020-12-15T00:00:00.0Z   
29046  2021-01-22T10:53:54.0Z  2021-01-22T11:06:01.0Z  2020-11-20T00:00:00.0Z   
29047  2021-01-22T10:54:55.0Z  2021-01-22T11:06:02.0Z  2020-11-20T00:00:00.0Z   

                 end_vod_date  
10     2021-12-31T23:59:59.0Z  
24     2021-02-28T23:59:00.0Z  
3187   2020-12-14T23:59:59.0Z  
3331   2020-12-14T23:59:59.0Z 

In [16]:
metadata[metadata['content_id']==948.0]

asset_id  content_id                    title        reduced_title  \
6421     11154       948.0  Un verano en pantalones  Un_Verano_En_Pantal   

            episode_title show_type  released_year country_of_origin category  \
6421  Un_Verano_En_Pantal       NaN         2005.0                US  Comedia   

                             keywords  ... language_rating dialog_rating  \
6421  Mujeres,De Libros,Amigos,Viajes  ...               N             N   

     fv_rating pay_per_view  pack_premium_1 pack_premium_2  \
6421         N            N               N              N   

                 create_date             modify_date          start_vod_date  \
6421  2020-12-29T18:10:46.0Z  2020-12-29T20:06:05.0Z  2020-12-02T00:00:00.0Z   

                end_vod_date  
6421  2021-01-01T23:59:00.0Z  

[1 rows x 30 columns]

In [17]:
metadata[metadata['content_id']==4315.0	]

asset_id  content_id                            title  \
32927     24509      4315.0  El día que el deporte se detuvo   

             reduced_title        episode_title show_type  released_year  \
32927  El_dia_que_el_depor  El_dia_que_el_depor       NaN         2021.0   

      country_of_origin            category  \
32927                US  Documental/Deporte   

                                      keywords  ... language_rating  \
32927  Pandemias,Deportistas,De Color,Sociedad  ...               N   

      dialog_rating fv_rating pay_per_view  pack_premium_1 pack_premium_2  \
32927             N         N            N               N              Y   

                  create_date             modify_date          start_vod_date  \
32927  2021-03-29T16:16:36.0Z  2021-03-29T17:06:04.0Z  2021-03-24T00:00:00.0Z   

                 end_vod_date  
32927  2022-01-10T23:59:00.0Z  

[1 rows x 30 columns]

Dado que el contenido id marca como 'Pelicula' al resto de los clasificados vamos a asignarle este valor.
y en el caso de l contenido 11154	y 24509 al ser unicos, vamos a tomar el mismo criterio.

In [18]:
metadata.loc[metadata['show_type'].isnull(),'show_type']='Película'

In [19]:
metadata.isnull().sum()

asset_id                  0
content_id                0
title                     0
reduced_title             0
episode_title             0
show_type                 0
released_year             0
country_of_origin         4
category                  0
keywords                  2
description               2
reduced_desc              0
cast_first_name        8711
credits_first_name    12533
run_time_min              0
audience                  1
made_for_tv               0
close_caption             0
sex_rating                0
violence_rating           0
language_rating           0
dialog_rating             0
fv_rating                 0
pay_per_view              0
pack_premium_1            0
pack_premium_2            0
create_date               0
modify_date               0
start_vod_date            0
end_vod_date              0
dtype: int64

###country_of_origin

In [20]:
metadata[metadata['country_of_origin'].isnull()]

asset_id  content_id                       title      reduced_title  \
7287      31314      1124.0           T:6 Ep:06 Dr. Pol      Dr_Pol_T6_E06   
7320      22004      1124.0           T:7 Ep:16 Dr. Pol      Dr_Pol_T7_E16   
7321       9452      1124.0           T:7 Ep:17 Dr. Pol      Dr_Pol_T7_E17   
17406      8946      2264.0  My Mother's Future Husband  My_Mothers_Future   

           episode_title show_type  released_year country_of_origin  \
7287     Dr. Franken-Pol        TV         2016.0               NaN   
7320    Llega la cigüeña        TV         2017.0               NaN   
7321       La inundación        TV         2017.0               NaN   
17406  My_Mothers_Future  Película         2014.0               NaN   

               category                               keywords  ...  \
7287   Reality/Animales     Veterinarios,Animales,Enfermedades  ...   
7320   Reality/Animales     Veterinarios,Animales,Enfermedades  ...   
7321   Reality/Animales     Veterinarios,Animales,Enfermedades  ...   
17406   Romance/Comedia  Melodrama,Tragicómica,Familia,Parejas  ...   

      language_rating dialog_rating fv_rating pay_per_view  pack_premium_1  \
7287                N             N         N            N               N   
7320                N             N         N            N               N   
7321                N             N         N            N               N   
17406               N             N         N            N               Y   

      pack_premium_2             create_date             modify_date  \
7287               N  2018-05-10T02:55:24.0Z  2019-03-26T12:40:02.0Z   
7320               N  2019-04-03T21:34:38.0Z  2019-10-17T11:26:23.0Z   
7321               N  2019-04-03T21:37:39.0Z  2019-10-17T11:26:24.0Z   
17406              N  2019-04-04T16:32:11.0Z  2019-10-17T11:26:26.0Z   

               start_vod_date            end_vod_date  
7287   2018-05-03T00:00:00.0Z  2021-01-16T23:59:59.0Z  
7320   2018-12-19T00:00:00.0Z  2021-01-16T23:59:00.0Z  
7321   2018-12-26T00:00:00.0Z  2021-01-16T23:59:00.0Z  
17406  2019-04-03T00:00:00.0Z  2022-03-31T23:59:00.0Z  

[4 rows x 30 columns]

In [21]:
metadata[metadata['content_id']==1124.0]

asset_id  content_id              title  reduced_title  \
7283      20070      1124.0  T:6 Ep:02 Dr. Pol   Dr_Pol_T6_E2   
7284       5900      1124.0  T:6 Ep:03 Dr. Pol   Dr_Pol_T6_E3   
7285      12651      1124.0  T:6 Ep:04 Dr. Pol   Dr_Pol_T6_E4   
7286       3304      1124.0  T:6 Ep:05 Dr. Pol   Dr_Pol_T6_E5   
7287      31314      1124.0  T:6 Ep:06 Dr. Pol  Dr_Pol_T6_E06   
7288      24603      1124.0  T:6 Ep:07 Dr. Pol  Dr_Pol_T6_E07   
7289       5512      1124.0  T:6 Ep:08 Dr. Pol  Dr_Pol_T6_E08   
7290      11866      1124.0  T:6 Ep:09 Dr. Pol  Dr_Pol_T6_E09   
7291      31286      1124.0  T:6 Ep:10 Dr. Pol  Dr_Pol_T6_E10   
7292      22327      1124.0  T:6 Ep:12 Dr. Pol  Dr_Pol_T6_E12   
7293       5307      1124.0  T:6 Ep:13 Dr. Pol  Dr_Pol_T6_E13   
7294       3294      1124.0  T:6 Ep:11 Dr. Pol  Dr_Pol_T6_E11   
7295      28693      1124.0  T:6 Ep:14 Dr. Pol  Dr_Pol_T6_E14   
7296      27618      1124.0  T:6 Ep:15 Dr. Pol  Dr_Pol_T6_E15   
7297      25314      1124.0  T:6 Ep:16 Dr. Pol  Dr_Pol_T6_E16   
7298      10338      1124.0  T:6 Ep:17 Dr. Pol  Dr_Pol_T6_E17   
7299      25084      1124.0  T:6 Ep:18 Dr. Pol   Dr_Pol_T6_18   
7300      14848      1124.0  T:7 Ep:01 Dr. Pol  Dr_Pol_T7_E01   
7301       4152      1124.0  T:7 Ep:02 Dr. Pol  Dr_Pol_T7_E02   
7302       6911      1124.0  T:7 Ep:03 Dr. Pol  Dr_Pol_T7_E03   
7303       2778      1124.0  T:6 Ep:01 Dr. Pol  Dr_Pol_T6_E01   
7304      10534      1124.0  T:6 Ep:19 Dr. Pol  Dr_Pol_T6_E19   
7305       4022      1124.0  T:7 Ep:04 Dr. Pol  Dr_Pol_T7_E04   
7306       1560      1124.0  T:7 Ep:05 Dr. Pol  Dr_Pol_T7_E05   
7307      13340      1124.0  T:7 Ep:06 Dr. Pol  Dr_Pol_T7_E06   
7308      19377      1124.0  T:7 Ep:07 Dr. Pol   Dr_Pol_T7_E7   
7309      21177      1124.0  T:7 Ep:08 Dr. Pol  Dr_Pol_T7_E08   
7310      21884      1124.0  T:7 Ep:09 Dr. Pol  Dr_Pol_T7_E09   
7311      31805      1124.0  T:7 Ep:10 Dr. Pol  Dr_Pol_T7_E10   
7312      32503      1124.0  T:7 Ep:11 Dr. Pol  Dr_Pol_T7_E11   
7313      21545      1124.0  T:7 Ep:12 Dr. Pol  Dr_Pol_T7_E12   
7314      31249      1124.0  T:7 Ep:13 Dr. Pol  Dr_Pol_T7_E13   
7315       9999      1124.0  T:7 Ep:14 Dr. Pol  Dr_Pol_T7_E14   
7316      16781      1124.0  T:7 Ep:15 Dr. Pol  Dr_Pol_T7_E15   
7317       1514      1124.0  T:7 Ep:18 Dr. Pol  Dr_Pol_T7_E18   
7318       7356      1124.0  T:7 Ep:19 Dr. Pol  Dr_Pol_T7_E19   
7319      18756      1124.0  T:7 Ep:20 Dr. Pol  Dr_Pol_T7_E20   
7320      22004      1124.0  T:7 Ep:16 Dr. Pol  Dr_Pol_T7_E16   
7321       9452      1124.0  T:7 Ep:17 Dr. Pol  Dr_Pol_T7_E17   
17697     16885      1124.0  T:8 Ep:01 Dr. Pol  Dr_Pol_T8_E01   
17794     24165      1124.0  T:8 Ep:02 Dr. Pol  Dr_Pol_T8_E02   
17877     16826      1124.0  T:8 Ep:03 Dr. Pol  Dr_Pol_T8_E03   
17959      9691      1124.0  T:8 Ep:04 Dr. Pol  Dr_Pol_T8_E04   
17999      5363      1124.0  T:8 Ep:05 Dr. Pol  Dr_Pol_T8_E05   
18035      5838      1124.0  T:8 Ep:06 Dr. Pol  Dr_Pol_T8_E06   
18112      8993      1124.0  T:8 Ep:08 Dr. Pol  Dr_Pol_T8_E08   

                           episode_title show_type  released_year  \
7283                      El apocalipsis        TV         2016.0   
7284         La Alegría de los Cachorros        TV         2016.0   
7285                    Dolorosa Puntada        TV         2016.0   
7286              El Sonido de una Cabra        TV         2016.0   
7287                     Dr. Franken-Pol        TV         2016.0   
7288       Día de Gracias con el Dr. Pol        TV         2016.0   
7289                       Un Gran Danés        TV         2016.0   
7290            Lecciones para el conejo        TV         2016.0   
7291                 Emergencia nocturna        TV         2016.0   
7292          El mejor doctor en narices        TV         2016.0   
7293                       Giro y hocico        TV         2016.0   
7294                       Patas mágicas        TV         2016.0   
7295      Grúas, esguinces y automóviles        TV    

In [22]:
metadata[metadata['content_id']==2264.0]

asset_id  content_id                       title      reduced_title  \
17406      8946      2264.0  My Mother's Future Husband  My_Mothers_Future   

           episode_title show_type  released_year country_of_origin  \
17406  My_Mothers_Future  Película         2014.0               NaN   

              category                               keywords  ...  \
17406  Romance/Comedia  Melodrama,Tragicómica,Familia,Parejas  ...   

      language_rating dialog_rating fv_rating pay_per_view  pack_premium_1  \
17406               N             N         N            N               Y   

      pack_premium_2             create_date             modify_date  \
17406              N  2019-04-04T16:32:11.0Z  2019-10-17T11:26:26.0Z   

               start_vod_date            end_vod_date  
17406  2019-04-03T00:00:00.0Z  2022-03-31T23:59:00.0Z  

[1 rows x 30 columns]

Al ver los content_id, presumimos que los contenidos son de US, menos el 8946 que investigando vemos que se trata de origen Canadiense.

In [23]:
metadata.loc[metadata['asset_id']==8946,'country_of_origin']='CA'

In [24]:
metadata.loc[metadata['country_of_origin'].isnull(),'country_of_origin']='US'

In [25]:
metadata.isnull().sum()

asset_id                  0
content_id                0
title                     0
reduced_title             0
episode_title             0
show_type                 0
released_year             0
country_of_origin         0
category                  0
keywords                  2
description               2
reduced_desc              0
cast_first_name        8711
credits_first_name    12533
run_time_min              0
audience                  1
made_for_tv               0
close_caption             0
sex_rating                0
violence_rating           0
language_rating           0
dialog_rating             0
fv_rating                 0
pay_per_view              0
pack_premium_1            0
pack_premium_2            0
create_date               0
modify_date               0
start_vod_date            0
end_vod_date              0
dtype: int64

In [26]:
metadata[metadata['keywords'].isnull()]

asset_id  content_id                                title  \
29761     25451      3914.0                 La supremacía Bourne   
30564     27065      1175.0  T:2 Ep:75 Drama total: La guardería   

             reduced_title        episode_title show_type  released_year  \
29761  La_supremacia_Bourn  La_supremacia_Bourn  Película         2004.0   
30564  Drama_total__T2_E75      Grody al máximo        TV         2019.0   

      country_of_origin                   category keywords  ...  \
29761                US            Acción/Suspenso      NaN  ...   
30564                US  Infantil/Dibujos Animados      NaN  ...   

      language_rating dialog_rating fv_rating pay_per_view  pack_premium_1  \
29761               N             N         N            N               Y   
30564               N             N         N            N               N   

      pack_premium_2             create_date             modify_date  \
29761              N  2021-02-09T12:40:00.0Z  2021-02-09T13:06:09.0Z   
30564              N  2021-03-01T23:10:13.0Z  2021-03-02T05:06:05.0Z   

               start_vod_date            end_vod_date  
29761  2021-02-07T00:00:00.0Z  2021-06-30T23:59:00.0Z  
30564  2021-02-20T00:00:00.0Z  2021-04-06T23:59:00.0Z  

[2 rows x 30 columns]

In [27]:
metadata[metadata['content_id']==1175.0]

asset_id  content_id                                title  \
7635      18103      1175.0  T:2 Ep:69 Drama total: La guardería   
7971      19846      1175.0  T:2 Ep:70 Drama total: La guardería   
8763        569      1175.0  T:2 Ep:71 Drama total: La guardería   
8781      10675      1175.0  T:2 Ep:53 Drama total: La guardería   
9152       5142      1175.0  T:2 Ep:72 Drama total: La guardería   
9205      24351      1175.0  T:2 Ep:54 Drama total: La guardería   
10077     29720      1175.0  T:2 Ep:56 Drama total: La guardería   
11403      3352      1175.0  T:2 Ep:57 Drama total: La guardería   
11709     18768      1175.0  T:2 Ep:58 Drama total: La guardería   
12381      8877      1175.0  T:2 Ep:59 Drama total: La guardería   
12382     14754      1175.0  T:2 Ep:60 Drama total: La guardería   
12697     24564      1175.0  T:2 Ep:73 Drama total: La guardería   
30148     15356      1175.0  T:2 Ep:74 Drama total: La guardería   
30564     27065      1175.0  T:2 Ep:75 Drama total: La guardería   
30826     12080      1175.0  T:2 Ep:76 Drama total: La guardería   
31765     14030      1175.0  T:2 Ep:77 Drama total: La guardería   

             reduced_title                     episode_title show_type  \
7635   Drama_total__T2_E69  Primos y rivales - Primera parte        TV   
7971   Drama_total__T2_E70  Primos y rivales - Segunda parte        TV   
8763   Drama_total__T2_E71              Por unos duncans más        TV   
8781   Drama_total__T2_E53                      Robo-maestra        TV   
9152   Drama_total__T2_E72                 Quiero ser payaso        TV   
9205   Drama_total__T2_E54         Guantes, guantes maestros        TV   
10077  Drama_total__T2_E56               El menti-ranosaurio        TV   
11403  Drama_total__T2_E57  Una idea Huev-xtremadamente mala        TV   
11709  Drama_total__T2_E58          Ejercitando los demonios        TV   
12381  Drama_total__T2_E59                  Pudín-pocalipsis        TV   
12382  Drama_total__T2_E60                     Mamá suplente        TV   
12697  Drama_total__T2_E73                    Somos juguetes        TV   
30148  Drama_total__T2_E74              Sueños probelmáticos        TV   
30564  Drama_total__T2_E75                   Grody al máximo        TV   
30826  Drama_total__T2_E76                     Un día peludo        TV   
31765  Drama_total__T2_E77               El revés del hambre        TV   

       released_year country_of_origin                   category  \
7635          2019.0                US  Infantil/Dibujos Animados   
7971          2019.0                US  Infantil/Dibujos Animados   
8763          2019.0                US  Infantil/Dibujos Animados   
8781          2019.0                US  Infantil/Dibujos Animados   
9152          2019.0                US  Infantil/Dibujos Animados   
9205          2019.0                US  Infantil/Dibujos Animados   
10077         2019.0                US  Infantil/Dibujos Animados   
11403         2019.0                US  Infantil/Dibujos Animados   
11709         2019.0                US  Infantil/Dibujos Animados   
12381         2019.0                US  Infantil/Dibujos Animados   
12382         2019.0                US  Infantil/Dibujos Animados   
12697         2019.0                US  Infantil/Dibujos Animados   
30148         2019.0                US  Infantil/Dibujos Animados   
30564         2019.0                US  Infantil/Dibujos Animados   
30826         2019.0                US  Infantil/Dibujos Animados   
31765         2019.0                US  Infantil/Dibujos Animados   

              keywords  ... language_rating dialog_rating fv_rating  \
7635        Travesuras  ...               N             N         N   
7971        Travesuras  ...               N             N         N   
8763        Travesuras  ...               N             N         N   
8781        Travesuras  ...               N             N         N   
9152        Travesuras  ...               N             N     

Vamos a usar como palabras clave 'Niños,Ingenioso' para el contenido 27065



In [28]:
metadata.loc[metadata['asset_id']==27065,'keywords']='Niños,Ingenioso'

In [29]:
metadata.loc[metadata['title']=='Bourne: El Ultimátum']['keywords']

29758    Agentes,Nueva York,Taquilleras,Espionaje
Name: keywords, dtype: object

Para el caso de 25451 vamos a copiar keywords de una pelicula de la misma saga, la 21097

In [30]:
metadata.loc[metadata['asset_id']==25451,'keywords']='Agentes,Nueva York,Taquilleras,Espionaje'

In [31]:
metadata.loc[metadata['asset_id']==25451,'keywords']

29761    Agentes,Nueva York,Taquilleras,Espionaje
Name: keywords, dtype: object

In [32]:
metadata.isnull().sum()

asset_id                  0
content_id                0
title                     0
reduced_title             0
episode_title             0
show_type                 0
released_year             0
country_of_origin         0
category                  0
keywords                  0
description               2
reduced_desc              0
cast_first_name        8711
credits_first_name    12533
run_time_min              0
audience                  1
made_for_tv               0
close_caption             0
sex_rating                0
violence_rating           0
language_rating           0
dialog_rating             0
fv_rating                 0
pay_per_view              0
pack_premium_1            0
pack_premium_2            0
create_date               0
modify_date               0
start_vod_date            0
end_vod_date              0
dtype: int64

###description

In [33]:
metadata[metadata['description'].isnull()]

asset_id  content_id              title   reduced_title episode_title  \
16238      5514      2168.0  T:1 Ep:03 Camping  Camping_T1_E03  Fishing Trip   
16330     21755      2168.0  T:1 Ep:04 Camping  Camping_T1_E04  Up All Night   

      show_type  released_year country_of_origin category  \
16238     Serie         2018.0                US  Comedia   
16330     Serie         2018.0                US  Comedia   

                                                keywords  ... language_rating  \
16238  Vacaciones,Parejas,Amigos,Optimistas,Fiestas,V...  ...               N   
16330  Vacaciones,Parejas,Amigos,Optimistas,Fiestas,V...  ...               N   

      dialog_rating fv_rating pay_per_view  pack_premium_1 pack_premium_2  \
16238             N         N            N               N              Y   
16330             N         N            N               N              Y   

                  create_date             modify_date          start_vod_date  \
16238  2018-11-20T16:33:34.0Z  2020-09-30T12:08:37.0Z  2018-11-19T00:00:00.0Z   
16330  2018-11-26T15:49:48.0Z  2020-09-30T12:11:40.0Z  2019-11-26T00:00:00.0Z   

                 end_vod_date  
16238  2021-03-31T23:59:59.0Z  
16330  2021-03-31T23:59:59.0Z  

[2 rows x 30 columns]

Para estos casos vamos a duplicar la descripcion reducida

In [34]:
metadata.loc[metadata['description'].isnull(),'description']=metadata[metadata['description'].isnull()]['reduced_desc']

In [35]:
metadata.loc[metadata['asset_id']==21755]

asset_id  content_id              title   reduced_title episode_title  \
16330     21755      2168.0  T:1 Ep:04 Camping  Camping_T1_E04  Up All Night   

      show_type  released_year country_of_origin category  \
16330     Serie         2018.0                US  Comedia   

                                                keywords  ... language_rating  \
16330  Vacaciones,Parejas,Amigos,Optimistas,Fiestas,V...  ...               N   

      dialog_rating fv_rating pay_per_view  pack_premium_1 pack_premium_2  \
16330             N         N            N               N              Y   

                  create_date             modify_date          start_vod_date  \
16330  2018-11-26T15:49:48.0Z  2020-09-30T12:11:40.0Z  2019-11-26T00:00:00.0Z   

                 end_vod_date  
16330  2021-03-31T23:59:59.0Z  

[1 rows x 30 columns]

In [36]:
metadata.isnull().sum()

asset_id                  0
content_id                0
title                     0
reduced_title             0
episode_title             0
show_type                 0
released_year             0
country_of_origin         0
category                  0
keywords                  0
description               0
reduced_desc              0
cast_first_name        8711
credits_first_name    12533
run_time_min              0
audience                  1
made_for_tv               0
close_caption             0
sex_rating                0
violence_rating           0
language_rating           0
dialog_rating             0
fv_rating                 0
pay_per_view              0
pack_premium_1            0
pack_premium_2            0
create_date               0
modify_date               0
start_vod_date            0
end_vod_date              0
dtype: int64

###cast_first_name

In [37]:
metadata[metadata['cast_first_name'].isnull()]

asset_id  content_id                                     title  \
9         32482         8.0  T:1 Ep:05 Mickey: Aventuras sobre ruedas   
11        30574        10.0        Las aventuras de Peabody y Sherman   
18        24600        17.0              Mune, el guardián de la luna   
19        14972        18.0     T:1 Ep:21 PJ Masks: Héroes en pijamas   
20        18155        19.0                   T:2 Ep:31 Patoaventuras   
...         ...         ...                                       ...   
33106     31039      1290.0                T:1 Ep:35 New Looney Tunes   
33107     16797      1290.0                T:1 Ep:39 New Looney Tunes   
33113     21955      1039.0               T:3 Ep:291 Hora de aventura   
33114      4823       154.0                  T:2 Ep:99 Phineas y Ferb   
33137     19624      4371.0          Zapatos rojos y los siete enanos   

             reduced_title                                      episode_title  \
9      Mickey_Avent_T1_E05     Agente cero, cero Goofy / El huevo exasperante   
11     Las_aventuras_de_Pe                                Las_aventuras_de_Pe   
18     Mune__el_guardian_d                                Mune__el_guardian_d   
19     PJ_Masks_Her_T1_E21  Catboy al cuadrado/ El súper instinto de Gecko...   
20     Patoaventura_T2_E31                                        Episodio 31   
...                    ...                                                ...   
33106  New_Looney_T_T1_E35                                  Bugs en el jardín   
33107  New_Looney_T_T1_E39                               La espía que me timó   
33113  Hora_de_ave_T3_E291                                    Demasiado joven   
33114   Phineas_Y_F_T2_E99                               Just Passing Through   
33137  Zapatos_rojos_y_los                                Zapatos_rojos_y_los   

      show_type  released_year country_of_origin                   category  \
9            TV         2017.0                US  Infantil/Dibujos Animados   
11     Película         2014.0                US         Infantil/Animación   
18     Película         2017.0                FR         Infantil/Animación   
19           TV         2017.0                FR  Infantil/Dibujos Animados   
20           TV         2017.0                US  Infantil/Dibujos Animados   
...         ...            ...               ...                        ...   
33106        TV         2015.0                US  Infantil/Dibujos Animados   
33107        TV         2015.0                US  Infantil/Dibujos Animados   
33113        TV         2011.0                US  Infantil/Dibujos Animados   
33114        TV         2007.0                US  Infantil/Dibujos Animados   
33137  Película         2019.0                US         Infantil/Animación   

                                                keywords  ... language_rating  \
9                                 Amigos,Aventuras,Autos  ...               N   
11     Dimensiones,Aventuras,Mascotas,Dreamworks,Familia  ...               N   
18                          Criaturas,Aventuras,Fantasía  ...               N   
19             Superhéroes,De Libros,Aventuras,Instituto  ...               N   
20                             Animales,Aventuras,Dinero  ...               N   
...                                                  ...  ...             ...   
33106                                 Animales,Ingenioso  ...               N   
33107                                 Animales,Ingenioso  ...               N   
33113                Ingenioso,Amigos,Mascotas,Aventuras  ...               N   
33114                                  Familia,Aventuras  ...               N   
33137               Aventuras,Criaturas,Princesas,Brujas  ...               N   

      dialog_rating fv_rating pay_per_view  pack_premium_1 pack_premium_2  \
9                 N         N            N               N              N   
11                N         N            Y               N              N   
18  

Lamentablemente esta columna la vamos a dejar sin tratar

In [38]:
metadata.isnull().sum()

asset_id                  0
content_id                0
title                     0
reduced_title             0
episode_title             0
show_type                 0
released_year             0
country_of_origin         0
category                  0
keywords                  0
description               0
reduced_desc              0
cast_first_name        8711
credits_first_name    12533
run_time_min              0
audience                  1
made_for_tv               0
close_caption             0
sex_rating                0
violence_rating           0
language_rating           0
dialog_rating             0
fv_rating                 0
pay_per_view              0
pack_premium_1            0
pack_premium_2            0
create_date               0
modify_date               0
start_vod_date            0
end_vod_date              0
dtype: int64

###credits_first_name

In [39]:
metadata[metadata['credits_first_name'].isnull()]

asset_id  content_id                                     title  \
7          2472         7.0     12/01 - Almorzando con Mirtha Legrand   
9         32482         8.0  T:1 Ep:05 Mickey: Aventuras sobre ruedas   
19        14972        18.0     T:1 Ep:21 PJ Masks: Héroes en pijamas   
20        18155        19.0                   T:2 Ep:31 Patoaventuras   
21        22721        20.0                  T:2 Ep:73 Puppy Dog Pals   
...         ...         ...                                       ...   
33110     24551      3891.0                            Ep:043 Züleyha   
33111     28170      1362.0                     Ep:050 Te doy la vida   
33112     27267      1409.0        03/29 - Corte y Confección Famosos   
33116     17733       933.0              T:10 Ep:95 El hormiguero 3.0   
33121     10134      4046.0   T:1 Ep:05 Momentos mágicos de la música   

             reduced_title                                      episode_title  \
7      Almorzando_ML_01-12                                  01 Diciembre 2019   
9      Mickey_Avent_T1_E05     Agente cero, cero Goofy / El huevo exasperante   
19     PJ_Masks_Her_T1_E21  Catboy al cuadrado/ El súper instinto de Gecko...   
20     Patoaventura_T2_E31                                        Episodio 31   
21     Puppy_Dog_Pa_T2_E73                                        Episodio 73   
...                    ...                                                ...   
33110         Zuleyha_E043                                        Episodio 43   
33111  Te_Doy_La_Vida_E050                                        Episodio 50   
33112  Corte_y_Confe_29-03                                      29 Marzo 2021   
33116  El_hormigue_T10_E95                                    Santi Rodríguez   
33121  Momentos_mag_T1_E05  Nace una estrella: Martha Argerich en Varsovia...   

      show_type  released_year country_of_origin                     category  \
7            TV         2019.0                AR  Interés General/Espectáculo   
9            TV         2017.0                US    Infantil/Dibujos Animados   
19           TV         2017.0                FR    Infantil/Dibujos Animados   
20           TV         2017.0                US    Infantil/Dibujos Animados   
21           TV         2017.0                US    Infantil/Dibujos Animados   
...         ...            ...               ...                          ...   
33110        TV         2020.0                TR                Drama/Romance   
33111        TV         2020.0                MX                        Drama   
33112        TV         2021.0                AR                 Reality/Moda   
33116        TV         2020.0                ES  Interés General/Entrevistas   
33121     Serie         2019.0                DE            Música/Documental   

                                                keywords  ... language_rating  \
7                                Actualidad,Espectáculos  ...               N   
9                                 Amigos,Aventuras,Autos  ...               N   
19             Superhéroes,De Libros,Aventuras,Instituto  ...               N   
20                             Animales,Aventuras,Dinero  ...               N   
21                                     Animales,Mascotas  ...               N   
...                                                  ...  ...             ...   
33110                           Parejas,Melodrama,Europa  ...               N   
33111                 Familia,Enfermedades,Latinoamérica  ...               N   
33112       Competencia,Belleza,Diseño,Moda,Celebridades  ...               N   
33116  Celebridades,Espectáculos,Actualidad,Stand Up,...  ...               N   
33121                            Música,Recital,Sinfonía  ...               N   

      dialog_rating fv_rating pay_per_view  pack_premium_1 pack_premium_2  \
7                 N         N            N               N              N   
9                 N         N            N               

Esta columna tambien queda sin tratar

In [40]:
metadata.isnull().sum()

asset_id                  0
content_id                0
title                     0
reduced_title             0
episode_title             0
show_type                 0
released_year             0
country_of_origin         0
category                  0
keywords                  0
description               0
reduced_desc              0
cast_first_name        8711
credits_first_name    12533
run_time_min              0
audience                  1
made_for_tv               0
close_caption             0
sex_rating                0
violence_rating           0
language_rating           0
dialog_rating             0
fv_rating                 0
pay_per_view              0
pack_premium_1            0
pack_premium_2            0
create_date               0
modify_date               0
start_vod_date            0
end_vod_date              0
dtype: int64

###audience

In [41]:
metadata[metadata['audience'].isnull()]

asset_id  content_id                        title        reduced_title  \
14523     26900      2064.0  T:1 Ep:08 Refugios Extremos  Refugios_Ext_T1_E08   

             episode_title show_type  released_year country_of_origin  \
14523  La cabaña giratoria        TV         2014.0                US   

                   category                      keywords  ...  \
14523  Reality/Restauración  Naturaleza,Decoración,Diseño  ...   

      language_rating dialog_rating fv_rating pay_per_view  pack_premium_1  \
14523               N             N         N            N               N   

      pack_premium_2             create_date             modify_date  \
14523              N  2018-01-17T17:51:55.0Z  2019-06-12T03:04:45.0Z   

               start_vod_date            end_vod_date  
14523  2018-01-05T00:00:00.0Z  2029-12-31T23:59:00.0Z  

[1 rows x 30 columns]

In [42]:
metadata[metadata['content_id']==2064.0]

asset_id  content_id                        title        reduced_title  \
14523     26900      2064.0  T:1 Ep:08 Refugios Extremos  Refugios_Ext_T1_E08   

             episode_title show_type  released_year country_of_origin  \
14523  La cabaña giratoria        TV         2014.0                US   

                   category                      keywords  ...  \
14523  Reality/Restauración  Naturaleza,Decoración,Diseño  ...   

      language_rating dialog_rating fv_rating pay_per_view  pack_premium_1  \
14523               N             N         N            N               N   

      pack_premium_2             create_date             modify_date  \
14523              N  2018-01-17T17:51:55.0Z  2019-06-12T03:04:45.0Z   

               start_vod_date            end_vod_date  
14523  2018-01-05T00:00:00.0Z  2029-12-31T23:59:00.0Z  

[1 rows x 30 columns]

In [43]:
metadata['audience'].unique()

array(['General', 'Juvenil', 'Preescolar', 'Teens', 'Hombres', 'Familiar',
       'Niños', 'Mujeres', 'Gaming', 'NIños', nan], dtype=object)

Este caso es muy particular, se trata del capitulo 8 de la temporada 1 de una serie que no tiene el resto de capitulos. Vamos a clasificarla como de audiencia general.

In [44]:
metadata.loc[metadata['audience'].isnull(),'audience']='General'

###Intento de meter data

In [45]:
metadata['category']=metadata["category"].str.split('/')

In [46]:
metadata.head()

asset_id  content_id                                  title  \
0     15188         0.0              Ep:17 Tiempos Compulsivos   
1     24940         1.0                                7 Cajas   
2     21939         2.0  La Maldición de las Hormigas Gigantes   
3      9005         3.0                   Una Mujer Fantástica   
4      7391         4.0                              Star Trek   

         reduced_title        episode_title show_type  released_year  \
0   Tiempos_Compul_E17          Episodio 17     Serie         2012.0   
1              7_Cajas              7_Cajas  Película         2012.0   
2  La_Maldicion_de_las  La_Maldicion_de_las  Película         2016.0   
3  Una_Mujer_Fantastic  Una_Mujer_Fantastic  Película         2017.0   
4            Star_Trek            Star_Trek  Película         2009.0   

  country_of_origin                     category  \
0                AR                      [Drama]   
1                PY           [Suspenso, Acción]   
2                FI            [Terror, Comedia]   
3                CL                      [Drama]   
4                US  [Ciencia Ficción, Aventura]   

                                            keywords  ... language_rating  \
0  Trastornos,Médicos,Tragicómica,Telenovela,Enfe...  ...               N   
1             Latinoamérica,Pobreza,Crimen,Pandillas  ...               N   
2  Criaturas,Plagas,Adolescentes,Fantasía,Video J...  ...               N   
3                         LGBT,Mujeres,Latinoamérica  ...               N   
4        Fantasía,Galaxia,Futurismo,Aliens,Criaturas  ...               N   

  dialog_rating fv_rating pay_per_view  pack_premium_1 pack_premium_2  \
0             N         N            N               N              N   
1             N         N            Y               N              N   
2             N         N            N               N              N   
3             N         N            N               Y              N   
4             N         N            Y               N              N   

              create_date             modify_date          start_vod_date  \
0  2017-12-01T10:18:15.0Z  2019-01-26T06:37:18.0Z  2017-12-01T00:00:00.0Z   
1  2017-12-19T20:58:15.0Z  2019-09-17T19:02:03.0Z  2017-12-15T00:00:00.0Z   
2  2018-02-16T13:51:07.0Z  2020-04-28T14:16:38.0Z  2018-01-25T00:00:00.0Z   
3  2018-05-26T11:58:44.0Z  2019-11-15T03:00:23.0Z  2018-05-27T00:00:00.0Z   
4  2019-05-03T20:07:24.0Z  2020-04-09T04:37:29.0Z  2019-05-02T00:00:00.0Z   

             end_vod_date  
0  2020-12-01T23:59:59.0Z  
1  2022-12-14T23:59:59.0Z  
2  2020-12-01T23:59:59.0Z  
3  2021-04-30T23:59:59.0Z  
4  2020-12-31T23:59:59.0Z  

[5 rows x 30 columns]

In [47]:
from sklearn.preprocessing import MultiLabelBinarizer

In [48]:
s = metadata['category']

In [49]:
mlb = MultiLabelBinarizer()

In [50]:
categorias=pd.DataFrame(mlb.fit_transform(s),columns=mlb.classes_, index=metadata.index)

In [51]:
metadata.join(categorias)

asset_id  content_id                                  title  \
0         15188         0.0              Ep:17 Tiempos Compulsivos   
1         24940         1.0                                7 Cajas   
2         21939         2.0  La Maldición de las Hormigas Gigantes   
3          9005         3.0                   Una Mujer Fantástica   
4          7391         4.0                              Star Trek   
...         ...         ...                                    ...   
33139      3683      1979.0              T:1 Ep:03 Allen V. Farrow   
33140     16810      1979.0              T:1 Ep:01 Allen V. Farrow   
33141     20468       943.0         T:2 Ep:34 Programa De Talentos   
33142      3779      1957.0                     T:2 Ep:05 Batwoman   
33143     30238      1041.0           T:1 Ep:25 Manzana y Cebollín   

             reduced_title        episode_title show_type  released_year  \
0       Tiempos_Compul_E17          Episodio 17     Serie         2012.0   
1                  7_Cajas              7_Cajas  Película         2012.0   
2      La_Maldicion_de_las  La_Maldicion_de_las  Película         2016.0   
3      Una_Mujer_Fantastic  Una_Mujer_Fantastic  Película         2017.0   
4                Star_Trek            Star_Trek  Película         2009.0   
...                    ...                  ...       ...            ...   
33139   Allen_V_Far_T1_E03           Episodio 3     Serie         2021.0   
33140   Allen_V_Far_T1_E01           Episodio 1     Serie         2021.0   
33141   Prog_De_Tal_T2_E34       El antagonista        TV         2011.0   
33142      Batwoman_T2_E05    Gore sobre lienzo     Serie         2021.0   
33143  Manzana_y_Ce_T1_E24  El corto de Manzana        TV         2017.0   

      country_of_origin                      category  \
0                    AR                       [Drama]   
1                    PY            [Suspenso, Acción]   
2                    FI             [Terror, Comedia]   
3                    CL                       [Drama]   
4                    US   [Ciencia Ficción, Aventura]   
...                 ...                           ...   
33139                US           [Drama, Documental]   
33140                US           [Drama, Documental]   
33141                US           [Infantil, Comedia]   
33142                US            [Acción, Aventura]   
33143                US  [Infantil, Dibujos Animados]   

                                                keywords  ... Político  \
0      Trastornos,Médicos,Tragicómica,Telenovela,Enfe...  ...        0   
1                 Latinoamérica,Pobreza,Crimen,Pandillas  ...        0   
2      Criaturas,Plagas,Adolescentes,Fantasía,Video J...  ...        0   
3                             LGBT,Mujeres,Latinoamérica  ...        0   
4            Fantasía,Galaxia,Futurismo,Aliens,Criaturas  ...        0   
...                                                  ...  ...      ...   
33139      Abusos,Familia,Juicio,Hechos Reales,Miniserie  ...        0   
33140      Abusos,Familia,Juicio,Hechos Reales,Miniserie  ...        0   
33141                            Instituto,Amigos,Música  ...        0   
33142               DC Comics,Superhéroes,Mujeres,Comics  ...        0   
33143                                  Amigos,Travesuras  ...        0   

      Reality Religión Restauración  Romance Suspenso Teatro Terror Viajes  \
0           0        0            0        0        0      0      0      0   
1           0        0            0        0        1      0      0      0   
2           0        0            0        0        0      0      1      0   
3           0        0            0        0        0      0      0      0   
4           0        0            0        0        0      0      0      0   
...       ...      ...          ...      ...      ...    ...    ...    ...   
33139       0        0            0        0        0      0      0      0   
33140       0        0            0        0        0  

In [52]:
metadata.to_csv('/content/drive/MyDrive/DATASCIENCE/Practico/Practico 3/metadatalimpia.csv', index=False)

##db_train

In [53]:
db_train.shape

(3657801, 7)

In [54]:
db_train.head()

customer_id  account_id device_type  asset_id                 tunein  \
0            0       90627  STATIONARY   18332.0  2021-02-18 22:52:00.0   
1            0       90627  STATIONARY   24727.0  2021-03-24 23:17:00.0   
2            1        3387         STB     895.0  2021-03-15 10:05:00.0   
3            1        3387         STB     895.0  2021-03-15 10:23:00.0   
4            1        3387         STB   26062.0  2021-03-16 09:24:00.0   

                 tuneout  resume  
0  2021-02-18 23:35:00.0       0  
1  2021-03-25 00:01:00.0       0  
2  2021-03-15 10:23:00.0       0  
3  2021-03-15 11:18:00.0       1  
4  2021-03-16 09:44:00.0       0

In [55]:
db_train.isnull().sum()

customer_id     0
account_id      0
device_type    29
asset_id       22
tunein          0
tuneout         0
resume          0
dtype: int64

###device_type

In [56]:
db_train[db_train['device_type'].isnull()]

customer_id  account_id device_type  asset_id                 tunein  \
528885         18693      105629         NaN   10939.0  2021-01-17 16:04:00.0   
666823         23439        6012         NaN    6371.0  2021-01-17 21:55:00.0   
737730         25729       46974         NaN    3481.0  2021-01-13 11:30:00.0   
737731         25729       46974         NaN    3481.0  2021-01-13 11:33:00.0   
737732         25729       46974         NaN    3481.0  2021-01-13 11:36:00.0   
931142         32196       52544         NaN     128.0  2021-02-10 23:36:00.0   
1020856        34976        6983         NaN   24467.0  2021-01-20 14:37:00.0   
1668627        53378       71627         NaN   21986.0  2021-01-13 06:36:00.0   
1785644        56361       74384         NaN   10208.0  2021-01-13 19:57:00.0   
2038544        64195       82302         NaN   32717.0  2021-01-16 17:53:00.0   
2038545        64195       82302         NaN   32717.0  2021-01-16 18:00:00.0   
2038546        64195       82302         NaN   32717.0  2021-01-16 19:42:00.0   
2038547        64195       82302         NaN   32717.0  2021-01-16 19:58:00.0   
2038548        64195       82302         NaN   22148.0  2021-01-16 20:26:00.0   
2038549        64195       82302         NaN   22148.0  2021-01-16 23:22:00.0   
2038550        64195       82302         NaN   11979.0  2021-01-17 09:55:00.0   
2038551        64195       82302         NaN   29845.0  2021-01-17 09:58:00.0   
2731969        82236      101874         NaN    9961.0  2021-01-15 16:12:00.0   
2740223        82422      102111         NaN   13909.0  2021-01-15 01:56:00.0   
2740225        82422      102111         NaN   25512.0  2021-01-15 02:57:00.0   
2993856        87982      108985         NaN   22305.0  2021-02-11 20:30:00.0   
3025633        88889      110083         NaN   27842.0  2021-02-10 22:08:00.0   
3025634        88889      110083         NaN    1542.0  2021-02-10 22:19:00.0   
3025635        88889      110083         NaN   18809.0  2021-02-10 22:52:00.0   
3144002        93905      100979         NaN    4239.0  2021-01-14 22:40:00.0   
3254238        97749       14892         NaN   30357.0  2021-01-16 01:22:00.0   
3254239        97749       14892         NaN   30357.0  2021-01-16 14:57:00.0   
3352578       101276       17621         NaN   13383.0  2021-01-16 21:54:00.0   
3635973       111507       25402         NaN   29565.0  2021-03-07 21:59:00.0   

                       tuneout  resume  
528885   2021-01-17 17:07:00.0       0  
666823   2021-01-17 21:58:00.0       0  
737730   2021-01-13 11:33:00.0       0  
737731   2021-01-13 11:36:00.0       0  
737732   2021-01-13 12:36:00.0       0  
931142   2021-02-10 23:43:00.0       0  
1020856  2021-01-20 15:13:00.0       0  
1668627  2021-01-13 06:38:00.0       0  
1785644  2021-01-13 20:47:00.0       0  
2038544  2021-01-16 17:56:00.0       0  
2038545  2021-01-16 18:33:00.0       0  
2038546  2021-01-16 19:55:00.0       0  
2038547  2021-01-16 20:05:00.0       0  
2038548  2021-01-16 20:59:00.0       0  
2038549  2021-01-16 23:42:00.0       0  
2038550  2021-01-17 09:58:00.0       0  
2038551  2021-01-17 10:45:00.0       0  
2731969  2021-01-15 17:40:00.0       0  
2740223  2021-01-15 02:46:00.0       0  
2740225  2021-01-15 03:50:00.0       0  
2993856  2021-02-11 20:37:00.0       0  
3025633  2021-02-10 22:14:00.0       0  
3025634  2021-02-10 22:22:00.0       0  
3025635  2021-02-10 23:04:00.0       0  
3144002  2021-01-14 22:49:00.0       0  
3254238  2021-01-16 01:29:00.0       0  
3254239  2021-01-16 15:03:00.0       0  
3352578  2021-01-16 22:24:00.0       0  
3635973  2021-03-07 22:52:00.0       0

Como no va a ser relevante el dispositivo desde el cual se conecta, vamos a ignorar estos nulos

###asset_id

In [57]:
db_train[db_train['asset_id'].isnull()]

customer_id  account_id   device_type  asset_id  \
106890          3746       28992           STB       NaN   
120175          4277       29416           STB       NaN   
146642          5174       30114           STB       NaN   
219038          7685       32112           STB       NaN   
249196          8809       33030         PHONE       NaN   
484894         17133       39757           STB       NaN   
726181         25327       46629         PHONE       NaN   
863794         29946       50616           STB       NaN   
1062022        36186       56086  CLOUD_CLIENT       NaN   
1238162        41180       60401           STB       NaN   
1321333        43598       62625           STB       NaN   
1652989        52885       71182           STB       NaN   
1854385        58326       76218           STB       NaN   
1860974        58550       76429           STB       NaN   
2200219        68671       86969           STB       NaN   
2348268        72746       91253           STB       NaN   
2982344        87692      108646  CLOUD_CLIENT       NaN   
3105229        92460       10784    STATIONARY       NaN   
3186447        95433       13058    STATIONARY       NaN   
3490722       106241       21345           STB       NaN   
3572923       109282       23657    STATIONARY       NaN   
3611436       110642       24742           STB       NaN   

                        tunein                tuneout  resume  
106890   2021-03-20 12:50:00.0  2021-03-20 12:52:00.0       1  
120175   2021-03-16 00:29:00.0  2021-03-16 00:36:00.0       0  
146642   2021-03-28 22:35:00.0  2021-03-28 22:39:00.0       0  
219038   2021-03-20 21:15:00.0  2021-03-20 21:55:00.0       0  
249196   2021-03-20 15:16:00.0  2021-03-20 15:18:00.0       0  
484894   2021-03-22 16:38:00.0  2021-03-22 16:51:00.0       0  
726181   2021-03-26 21:59:00.0  2021-03-26 22:09:00.0       0  
863794   2021-03-24 22:07:00.0  2021-03-24 22:09:00.0       0  
1062022  2021-03-17 18:56:00.0  2021-03-17 18:58:00.0       0  
1238162  2021-03-12 17:47:00.0  2021-03-12 17:52:00.0       0  
1321333  2021-03-23 12:41:00.0  2021-03-23 12:45:00.0       0  
1652989  2021-03-21 01:00:00.0  2021-03-21 01:03:00.0       0  
1854385  2021-03-13 13:18:00.0  2021-03-13 13:21:00.0       0  
1860974  2021-03-17 19:24:00.0  2021-03-17 19:30:00.0       0  
2200219  2021-03-25 17:33:00.0  2021-03-25 18:36:00.0       0  
2348268  2021-03-27 03:32:00.0  2021-03-27 03:38:00.0       0  
2982344  2021-03-17 23:55:00.0  2021-03-18 00:07:00.0       0  
3105229  2021-03-12 13:32:00.0  2021-03-12 13:36:00.0       0  
3186447  2021-03-15 20:22:00.0  2021-03-15 20:26:00.0       0  
3490722  2021-03-23 15:26:00.0  2021-03-23 15:58:00.0       0  
3572923  2021-03-20 01:43:00.0  2021-03-20 01:46:00.0       0  
3611436  2021-03-18 21:10:00.0  2021-03-18 21:14:00.0       1

En este caso, dado que no tenemos el contenido que se vió, vamos a proceder a eliminar estas filas.

In [58]:
db_train=db_train.drop(db_train[db_train['asset_id'].isnull()].index,axis=0)

In [ ]:
db_train.isnull().sum()

#TRAINT-TEST

Para este ejercicio solo voy a tomar contenido visto como aquel en el que quedó registrado en la columna asset_id, omitiendo si la visualizacion del contenido fue casi nula, parcial o total.
 Es por esto que solo me voy a quedar con las columnas account_id, asset_id.
 Tambien para su posterior separacion en train y test vamos a conservar la columna tuneout.

In [ ]:
db=db_train[['account_id','asset_id','tuneout']]

In [ ]:
db.shape

In [ ]:
db.head()

Lo siguiente es transformar asset_id en content_id

In [ ]:
db_mod=pd.merge(db,metadata[['asset_id','content_id','end_vod_date','category','keywords']],how='left')

In [ ]:
db_mod.head()

In [ ]:
db_mod=db_mod.drop(['asset_id'],axis=1)

In [ ]:
db_mod.head()

##Cambio de columna Time


In [ ]:
db_mod['end_vod_date']=pd.to_datetime(db_mod.end_vod_date)

In [ ]:
db_mod['tuneout']=pd.to_datetime(db_mod.tuneout)

In [ ]:
db_mod['end_vod_date'] = db_mod['end_vod_date'].apply(lambda x: x.replace(tzinfo=None).date())

In [ ]:
db_mod['tuneout'] = db_mod['tuneout'].apply(lambda x: x.replace(tzinfo=None).date())

##Guardamos el archivo

In [ ]:
db_mod.to_csv('/content/drive/MyDrive/DATASCIENCE/Practico/Practico 3/db_mod.csv', index=False)

##Quitamos contenido.
Vamos a eliminar todo el contenido que ya no a va a estar disponible.

In [ ]:
db_mod.shape

In [ ]:
type(db_mod['end_vod_date'][0])

In [ ]:
db_mod=db_mod[db_mod['end_vod_date']>datetime(2021, 3, 1).date()]

In [ ]:
db_mod.shape

In [ ]:
min(db_mod['end_vod_date'])

In [ ]:
db_mod=db_mod.drop('end_vod_date', axis=1)

In [ ]:
db_mod

##Separacion entre train y test
Para esta separacion el criterio es por fecha, todo lo anterior a 1 de marzo de 2021 será train y lo posterior test.

In [ ]:
db_train=db_mod[db_mod['tuneout']<=datetime(2021, 3, 1).date()]

In [ ]:
db_test=db_mod[db_mod['tuneout']>datetime(2021, 3, 1).date()]

In [ ]:
db_train.shape

In [ ]:
db_test.shape

##Eliminamos tuneout

In [ ]:
db_train=db_train.drop('tuneout',axis=1)

In [ ]:
db_test=db_test.drop('tuneout',axis=1)

In [ ]:
db_train.head()

In [ ]:
db_test.head()

##Eliminamos duplicados
Vamos a eliminar los registros duplicados, estos registros se generaron dado que al unificar los capitulos por serie, un id que vio varios capitulos de una serie, va a figurar varias veces.

In [ ]:
db_train.shape

In [ ]:
db_train=db_train.drop_duplicates(subset=['account_id', 'content_id'], keep='last')

In [ ]:
db_train.shape

In [ ]:
db_train['account_id'].nunique()

In [ ]:
db_test=db_test.drop_duplicates()

#Contenido mas popular

Aca vamos a colocar en la variable 'popular' el contenido mas visto. Luego recomendaremos estos contenidos a los usuarios nuevos en la funcion Recomendar.

In [ ]:
populares=db_train['content_id'].value_counts().head(20).index.to_list()

In [ ]:
populares

#Matriz Interaction 


In [ ]:
db_train.head()

In [ ]:
ddbb=db_train.assign(miro=1)

In [ ]:
interaction=ddbb.pivot(index='account_id',columns='content_id', values='miro')

In [ ]:
interaction.shape

In [ ]:
interaction.fillna(value=0,inplace=True)

In [ ]:
interaction

##indices de referencia

In [ ]:
account_id = list(interaction.index)
user_dict = {}
counter = 0 
for i in account_id:
    user_dict[i] = counter
    counter += 1

In [ ]:
content_id = list(interaction.columns)
item_dict = {}
counter = 0 
for i in content_id:
    item_dict[i] = counter
    counter += 1

##Converimos la matriz interaction

In [ ]:
from scipy.sparse import csr_matrix

In [ ]:
interaction_csr = csr_matrix(interaction.values)
interaction_csr

In [ ]:
interaction_csr.shape

In [ ]:
categorias2= csr_matrix(categorias.values)

#ENTRENAMOS EL MODELO

In [ ]:
pip install lightfm

In [ ]:
from lightfm import LightFM

In [ ]:
model = LightFM(loss='warp',no_components=10)


In [ ]:
model=model.fit(interaction_csr,epochs=20,item_features=categorias2)

In [ ]:
user_x = user_dict[18]
n_users, n_items = interaction.shape
item_ids = np.arange(n_items)
#ACA GENERAMOS PREDICCIONES
preds = model.predict(user_ids=user_x, item_ids = item_ids)

In [ ]:
preds

In [ ]:
scores = pd.Series(preds)
scores.index = interaction.columns
recomms_18 = list(pd.Series(scores.sort_values(ascending=False).index))[:10] #obtenemos las primeras 10

In [ ]:
recomms_18

#Funcion Recomendar

In [ ]:
def Recomendador(Unser_ID):
  #Esta funcion recibe un usuario id, lo busca en la base de datos y si lo encuentra le da una recomendacion de 40 contenidos recomendados,
  #caso contrario recomienda los 20 contenidos mas populares.
  
  if Unser_ID in list(interaction.index):
    
    user_x = user_dict[Unser_ID]
    n_users, n_items = interaction.shape
    item_ids = np.arange(n_items)
    preds = model.predict(user_ids=user_x, item_ids = item_ids)

    scores = pd.Series(preds)
    scores.index = interaction.columns
    Recomendadas = list(pd.Series(scores.sort_values(ascending=False).index))[:20] #obtenemos las primeras 10

    #chequeamos que el contenido no fue previamente visto
    vio=db_train[db_train['account_id']==2].content_id.unique()
    for i in vio:
      for j in Recomendadas:
        if i==j:
          Recomendadas.remove(i)


  else:
    Recomendadas=populares

  return Recomendadas

In [ ]:
Recomendador(2)

#Evaluacion

In [ ]:
tabla_recomendada=db_test[['account_id','content_id']]

In [ ]:
tabla_recomendada.head()

In [ ]:
tabla_recomendada.shape

In [ ]:
tabla_recomendada['account_id'].nunique()

In [ ]:
tabla_recomendada=tabla_recomendada.groupby('account_id', as_index=False).agg({"content_id": "unique"})

In [ ]:
tabla_recomendada = pd.DataFrame(tabla_recomendada)

In [ ]:
tabla_recomendada.head()

In [ ]:
tabla_recomendada.shape

In [ ]:
Rec=[] #Genero una columna con todas recomendaciones populares.
for i in tabla_recomendada['account_id']:
  Rec.append(populares)

In [ ]:
tabla_recomendada.shape

In [ ]:
tabla_recomendada['Popular']=Rec

In [ ]:
tabla_recomendada.head()

In [ ]:
tabla_recomendada = pd.DataFrame(tabla_recomendada)

In [ ]:
aps = [] # lista vacía para ir almacenando la AP de cada recomendación


for pred, label in tabla_recomendada[["content_id", "Popular"]].values:
  n = len(pred) # cantidad de elementos recomendados
  arange = np.arange(n, dtype=np.int32) + 1. # indexamos en base 1 
  rel_k = np.in1d(pred[:n], label) # lista de booleanos que indican la relevancia de cada ítem
  tp = np.ones(rel_k.sum(), dtype=np.int32).cumsum() # lista con el contador de verdaderos positivos
  denom = arange[rel_k] # posiciones donde se encuentran los ítems relantes
  ap = (tp / denom).sum() / len(label) # average precision
  aps.append(ap)

In [ ]:
MAP = np.mean(aps)
print(f'mean average precision = {round(MAP, 5)}')

In [ ]:
MiRec=[]
for i in tabla_recomendada['account_id']:
  recomendaciones=Recomendador(i)
  MiRec.append(recomendaciones)


In [ ]:
len(MiRec)

In [ ]:
tablafinal=tabla_recomendada.copy()

In [ ]:
tablafinal['Mi_recomendacion']=MiRec

In [ ]:
aps = [] # lista vacía para ir almacenando la AP de cada recomendación


for pred, label in tablafinal[["content_id", 'Mi_recomendacion']].values:
  n = len(pred) # cantidad de elementos recomendados
  arange = np.arange(n, dtype=np.int32) + 1. # indexamos en base 1 
  rel_k = np.in1d(pred[:n], label) # lista de booleanos que indican la relevancia de cada ítem
  tp = np.ones(rel_k.sum(), dtype=np.int32).cumsum() # lista con el contador de verdaderos positivos
  denom = arange[rel_k] # posiciones donde se encuentran los ítems relantes
  ap = (tp / denom).sum() / len(label) # average precision
  aps.append(ap)

In [ ]:
MAP = np.mean(aps)
print(f'mean average precision = {round(MAP, 5)}')